src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,1229 @@
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * alter_table_set_access_method gets a distributed table and an access
+ * method and changes table's access method into that.
+ */
+Datum
+alter_table_set_access_method(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+
+	text *accessMethodText = PG_GETARG_TEXT_P(1);
+	char *accessMethod = text_to_cstring(accessMethodText);
+
+	CheckCitusVersion(ERROR);
+	EnsureRelationExists(relationId);
+	EnsureTableOwner(relationId);
+
+	if (IsCitusTable(relationId))
+	{
+		EnsureCoordinator();
+	}
+
+	TableConversionParameters params = {
+		.relationId = relationId,
+		.accessMethod = accessMethod
+	};
+
+	AlterTableSetAccessMethod(&params);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * worker_change_sequence_dependency is a wrapper UDF for
+ * changeDependencyFor function
+ */
+Datum
+worker_change_sequence_dependency(PG_FUNCTION_ARGS)
+{
+	Oid sequenceOid = PG_GETARG_OID(0);
+	Oid sourceRelationOid = PG_GETARG_OID(1);
+	Oid targetRelationOid = PG_GETARG_OID(2);
+
+	changeDependencyFor(RelationRelationId, sequenceOid,
+						RelationRelationId, sourceRelationOid, targetRelationOid);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * UndistributeTable undistributes the given table. It uses ConvertTable function to
+ * create a new local table and move everything to that table.
+ *
+ * The local tables, tables with references, partition tables and foreign tables are
+ * not supported. The function gives errors in these cases.
+ */
+TableConversionReturn *
+UndistributeTable(TableConversionParameters *params)
+{
+	if (!IsCitusTable(params->relationId))
+	{
+		ereport(ERROR, (errmsg("cannot undistribute table "
+							   "because the table is not distributed")));
+	}
+
+	EnsureTableNotReferencing(params->relationId);
+	EnsureTableNotReferenced(params->relationId);
+	EnsureTableNotForeign(params->relationId);
+	EnsureTableNotPartition(params->relationId);
+
+	params->conversionType = UNDISTRIBUTE_TABLE;
+	params->shardCountIsNull = true;
+	TableConversionState *con = CreateTableConversion(params);
+	return ConvertTable(con);
+}
+
+
+/*
+ * AlterDistributedTable changes some properties of the given table. It uses
+ * ConvertTable function to create a new local table and move everything to that table.
+ *
+ * The local and reference tables, tables with references, partition tables and foreign
+ * tables are not supported. The function gives errors in these cases.
+ */
+TableConversionReturn *
+AlterDistributedTable(TableConversionParameters *params)
+{
+	if (!IsCitusTableType(params->relationId, DISTRIBUTED_TABLE))
+	{
+		ereport(ERROR, (errmsg("cannot alter table because the table "
+							   "is not distributed")));
+	}
+
+	EnsureTableNotForeign(params->relationId);
+	EnsureTableNotPartition(params->relationId);
+
+	params->conversionType = ALTER_DISTRIBUTED_TABLE;
+	TableConversionState *con = CreateTableConversion(params);
+	CheckAlterDistributedTableConversionParameters(con);
+
+	SetLocalMultiShardModifyModeToSequential();
```

Comment:
This slows things down quite significantly, because we lose multi-core parallelism. It should only be necessary when any of the co-located tables have a foreign key to a reference table.